File: windows_app_winui_beta/src/SaveStories.WinUI.Beta/Services/NotionRoutingRulesSource.cs
using System.Text.Json;
using System.Text.RegularExpressions;

namespace SaveMe.WinUI.Beta.Services;

public sealed partial class NotionRoutingRulesSource
{
    public const string DefaultPageUrl = "https://narrow-park-cda.notion.site/366f65c3de6780e8bb03f4bdda65f5f8";

    private static readonly HttpClient Http = new();

    public async Task<string> FetchRulesAsync(
        string pageUrl = DefaultPageUrl,
        CancellationToken cancellationToken = default)
    {
        var pageId = PageIdFromUrl(pageUrl);
        var endpoint = ApiEndpointFromUrl(pageUrl);
        object cursor = new Dictionary<string, object?> { ["stack"] = Array.Empty<object>() };
        var chunkNumber = 0;
        var chunks = new List<string>();

        while (chunkNumber < 10)
        {
            var payload = new Dictionary<string, object?>
            {
                ["pageId"] = pageId,
                ["limit"] = 100,
                ["cursor"] = cursor,
                ["chunkNumber"] = chunkNumber,
                ["verticalColumns"] = false,
            };
            using var request = new HttpRequestMessage(HttpMethod.Post, endpoint)
            {
                Content = new StringContent(JsonSerializer.Serialize(payload), System.Text.Encoding.UTF8, "application/json"),
            };
            request.Headers.UserAgent.ParseAdd("SaveMe-WinUI-NotionRoutingRulesSource");

            using var response = await Http.SendAsync(request, cancellationToken);
            response.EnsureSuccessStatusCode();
            await using var stream = await response.Content.ReadAsStreamAsync(cancellationToken);
            using var document = await JsonDocument.ParseAsync(stream, cancellationToken: cancellationToken);

            var rules = ParseRules(document.RootElement);
            if (!string.IsNullOrWhiteSpace(rules))
            {
                chunks.Add(rules);
            }

            if (!TryGetNextCursor(document.RootElement, out cursor))
            {
                break;
            }

            chunkNumber++;
        }

        var lines = string.Join(Environment.NewLine, chunks)
            .Split(Environment.NewLine, StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries)
            .Distinct(StringComparer.OrdinalIgnoreCase);
        return string.Join(Environment.NewLine, lines);
    }

    public static string ParseRules(JsonElement root)
    {
        if (!root.TryGetProperty("recordMap", out var recordMap)
            || !recordMap.TryGetProperty("block", out var blocks)
            || blocks.ValueKind != JsonValueKind.Object)
        {
            return "";
        }

        var texts = new List<string>();
        foreach (var block in blocks.EnumerateObject())
        {
            if (block.Value.TryGetProperty("value", out var valueContainer)
                && valueContainer.TryGetProperty("value", out var value)
                && value.TryGetProperty("properties", out var properties)
                && properties.TryGetProperty("title", out var title))
            {
                texts.Add(NotionRichText(title));
            }
        }

        return NormalizeRules(string.Join(Environment.NewLine, texts));
    }

    public static string NormalizeRules(string text)
    {
        var rules = new List<string>();
        foreach (var rawLine in text.Split('\n', StringSplitOptions.TrimEntries))
        {
            var line = rawLine.Trim();
            if (line.Length == 0
                || line.Contains("Правила для сортировки", StringComparison.OrdinalIgnoreCase))
            {
                continue;
            }

            var separator = line.IndexOf('=');
            if (separator < 0)
            {
                continue;
            }

            var left = line[..separator];
            var target = line[(separator + 1)..].Trim();
            if (target.Length == 0)
            {
                continue;
            }

            foreach (var username in NormalizedUsernames(left))
            {
                rules.Add($"{username} = {target}");
            }
        }

        return string.Join(Environment.NewLine, rules);
    }

    private static IEnumerable<string> NormalizedUsernames(string value)
    {
        foreach (var rawPart in value.Split(':', StringSplitOptions.TrimEntries | StringSplitOptions.RemoveEmptyEntries))
        {
            var part = ParentheticalRegex().Replace(rawPart, "");
            var cleaned = part.Trim().Trim('@', '*', '/', '\\', '.', ',', ';', ':', '!', '?', '[', ']', '{', '}', '<', '>', '"', '\'', '`');
            if (cleaned.Length > 0)
            {
                yield return cleaned;
            }
        }
    }

    private static string NotionRichText(JsonElement value)
    {
        if (value.ValueKind == JsonValueKind.String)
        {
            return value.GetString() ?? "";
        }

        if (value.ValueKind != JsonValueKind.Array)
        {
            return "";
        }

        var parts = new List<string>();
        foreach (var fragment in value.EnumerateArray())
        {
            if (fragment.ValueKind == JsonValueKind.String)
            {
                parts.Add(fragment.GetString() ?? "");
            }
            else if (fragment.ValueKind == JsonValueKind.Array && fragment.GetArrayLength() > 0)
            {
                var text = fragment[0];
                if (text.ValueKind == JsonValueKind.String)
                {
                    parts.Add(text.GetString() ?? "");
                }
            }
        }

        return string.Concat(parts);
    }

    private static bool TryGetNextCursor(JsonElement root, out object cursor)
    {
        cursor = new Dictionary<string, object?>();
        if (!root.TryGetProperty("cursor", out var cursorElement)
            || cursorElement.ValueKind != JsonValueKind.Object
            || !cursorElement.TryGetProperty("stack", out var stack)
            || stack.ValueKind != JsonValueKind.Array
            || stack.GetArrayLength() == 0)
        {
            return false;
        }

        cursor = JsonSerializer.Deserialize<object>(cursorElement.GetRawText()) ?? new Dictionary<string, object?>();
        return true;
    }

    private static string PageIdFromUrl(string pageUrl)
    {
        var match = Regex.Match(pageUrl, "[0-9a-fA-F]{32}");
        if (!match.Success)
        {
            throw new InvalidOperationException("Некорректная ссылка на Notion-страницу с правилами.");
        }

        var compact = match.Value.ToLowerInvariant();
        return string.Join(
            "-",
            compact[..8],
            compact[8..12],
            compact[12..16],
            compact[16..20],
            compact[20..]);
    }

    private static string ApiEndpointFromUrl(string pageUrl)
    {
        var uri = new Uri(pageUrl);
        return $"{uri.Scheme}://{uri.Host}/api/v3/loadCachedPageChunk";
    }

    [GeneratedRegex(@"\s*\([^)]*\)")]
    private static partial Regex ParentheticalRegex();
}

File: Sources/SaveStories/Services/NotionRoutingRulesSource.swift
import Foundation

struct NotionRoutingRulesSource {
    static let defaultPageURL = URL(string: "https://narrow-park-cda.notion.site/366f65c3de6780e8bb03f4bdda65f5f8")!

    private let session: URLSession

    init(session: URLSession = .shared) {
        self.session = session
    }

    func fetchRules(from pageURL: URL = Self.defaultPageURL) async throws -> String {
        let pageID = try Self.pageID(from: pageURL)
        let endpoint = try Self.apiEndpoint(for: pageURL)
        var cursor: [String: Any] = ["stack": []]
        var chunkNumber = 0
        var rules: [String] = []
        var seen = Set<String>()

        while chunkNumber < 10 {
            let payload: [String: Any] = [
                "pageId": pageID,
                "limit": 100,
                "cursor": cursor,
                "chunkNumber": chunkNumber,
                "verticalColumns": false,
            ]

            var request = URLRequest(url: endpoint)
            request.httpMethod = "POST"
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
            request.setValue("SaveMe/NotionRoutingRulesSource", forHTTPHeaderField: "User-Agent")
            request.httpBody = try JSONSerialization.data(withJSONObject: payload)

            let (data, response) = try await session.data(for: request)
            if let httpResponse = response as? HTTPURLResponse,
               !(200...299).contains(httpResponse.statusCode) {
                throw NotionRoutingRulesSourceError.httpStatus(httpResponse.statusCode)
            }

            for rule in try Self.parseRules(from: data).split(whereSeparator: \.isNewline).map(String.init)
                where seen.insert(rule.lowercased()).inserted {
                rules.append(rule)
            }

            guard let nextCursor = try Self.nextCursor(from: data),
                  let stack = nextCursor["stack"] as? [Any],
                  !stack.isEmpty else {
                break
            }

            cursor = nextCursor
            chunkNumber += 1
        }

        return rules.joined(separator: "\n")
    }

    static func parseRules(from data: Data) throws -> String {
        let object = try JSONSerialization.jsonObject(with: data)
        guard let root = object as? [String: Any],
              let recordMap = root["recordMap"] as? [String: Any],
              let blocks = recordMap["block"] as? [String: Any] else {
            return ""
        }

        var texts: [String] = []
        for block in blocks.values {
            guard let blockContainer = block as? [String: Any],
                  let valueContainer = blockContainer["value"] as? [String: Any],
                  let value = valueContainer["value"] as? [String: Any],
                  let properties = value["properties"] as? [String: Any],
                  let title = properties["title"] else {
                continue
            }

            texts.append(notionRichText(title))
        }

        return normalizeRules(texts.joined(separator: "\n"))
    }

    static func normalizeRules(_ text: String) -> String {
        var rules: [String] = []
        for rawLine in text.components(separatedBy: .newlines) {
            let line = rawLine.trimmingCharacters(in: .whitespacesAndNewlines)
            guard !line.isEmpty,
                  !line.localizedCaseInsensitiveContains("Правила для сортировки"),
                  let separator = line.firstIndex(of: "=") else {
                continue
            }

            let left = String(line[..<separator])
            let target = String(line[line.index(after: separator)...])
                .trimmingCharacters(in: .whitespacesAndNewlines)
            guard !target.isEmpty else { continue }

            for username in normalizedUsernames(from: left) {
                rules.append("\(username) = \(target)")
            }
        }

        return rules.joined(separator: "\n")
    }

    private static func normalizedUsernames(from value: String) -> [String] {
        value
            .split(separator: ":")
            .compactMap { rawPart in
                let part = String(rawPart)
                    .replacingOccurrences(of: #"\s*\([^)]*\)"#, with: "", options: .regularExpression)
                let cleaned = part
                    .trimmingCharacters(in: .whitespacesAndNewlines)
                    .trimmingCharacters(in: CharacterSet(charactersIn: "@*/\\.,;:!?[]{}<>\"'`"))
                return cleaned.isEmpty ? nil : cleaned
            }
    }

    private static func notionRichText(_ value: Any) -> String {
        guard let fragments = value as? [Any] else {
            return value as? String ?? ""
        }

        return fragments.compactMap { fragment in
            if let tuple = fragment as? [Any],
               let text = tuple.first as? String {
                return text
            }
            return fragment as? String
        }.joined()
    }

    private static func pageID(from url: URL) throws -> String {
        let source = url.absoluteString
        guard let regex = try? NSRegularExpression(pattern: #"[0-9a-fA-F]{32}"#),
              let match = regex.firstMatch(in: source, range: NSRange(source.startIndex..<source.endIndex, in: source)),
              let range = Range(match.range, in: source) else {
            throw NotionRoutingRulesSourceError.invalidPageURL
        }

        let compact = String(source[range]).lowercased()
        return [
            compact.prefix(8),
            compact.dropFirst(8).prefix(4),
            compact.dropFirst(12).prefix(4),
            compact.dropFirst(16).prefix(4),
            compact.dropFirst(20),
        ].map(String.init).joined(separator: "-")
    }

    private static func apiEndpoint(for pageURL: URL) throws -> URL {
        guard let scheme = pageURL.scheme,
              let host = pageURL.host,
              let endpoint = URL(string: "\(scheme)://\(host)/api/v3/loadCachedPageChunk") else {
            throw NotionRoutingRulesSourceError.invalidPageURL
        }
        return endpoint
    }

    private static func nextCursor(from data: Data) throws -> [String: Any]? {
        let object = try JSONSerialization.jsonObject(with: data)
        guard let root = object as? [String: Any] else { return nil }
        return root["cursor"] as? [String: Any]
    }
}

enum NotionRoutingRulesSourceError: LocalizedError {
    case invalidPageURL
    case httpStatus(Int)

    var errorDescription: String? {
        switch self {
        case .invalidPageURL:
            "Некорректная ссылка на Notion-страницу с правилами."
        case let .httpStatus(status):
            "Notion вернул HTTP \(status)."
        }
    }
}
